gcp/src/main/java/org/apache/iceberg/gcp/biglake/BigLakeTableOperations.java

```diff
@@ -0,0 +1,227 @@
+
+  @Override
+  public String tableName() {
+    return String.format(
+        "%s.%s.%s", tableName.getCatalog(), tableName.getDatabase(), tableName.getTable());
+  }
+
+  @Override
+  public FileIO io() {
+    return fileIO;
+  }
+
+  private void createTable(String newMetadataLocation, TableMetadata metadata) {
+    LOG.debug("Creating a new Iceberg table: {}", tableName());
+    client.createTable(tableName, makeNewTable(metadata, newMetadataLocation));
+  }
+
+  /** Update table properties with concurrent update detection using etag. */
+  private void updateTable(
+      String oldMetadataLocation, String newMetadataLocation, TableMetadata metadata) {
+    Table table = client.getTable(tableName);
+    String etag = table.getEtag();
+    Preconditions.checkArgument(
+        !etag.isEmpty(),
+        "Etag of legacy table %s is empty, manually update the table by BigLake API or recreate and retry",
+        tableName());
+    HiveTableOptions options = table.getHiveOptions();
+
+    // If `metadataLocationFromMetastore` is different from metadata location of base, it means
+    // someone has updated metadata location in metastore, which is a conflict update.
+    String metadataLocationFromMetastore =
+        options.getParametersOrDefault(METADATA_LOCATION_PROP, "");
+    if (!metadataLocationFromMetastore.isEmpty()
+        && !metadataLocationFromMetastore.equals(oldMetadataLocation)) {
+      throw new CommitFailedException(
+          "Base metadata location '%s' is not same as the current table metadata location '%s' for"
+              + " %s.%s",
+          oldMetadataLocation,
+          metadataLocationFromMetastore,
+          tableName.getDatabase(),
+          tableName.getTable());
+    }
+
+    try {
+      // Updating a BLMS table with etag. The BLMS server transactionally (1) checks that the etag
+      // of a table on server is the same as the etag provided by the client, and (2) updates the
+      // table (and its etag). The server returns an error containing message "etag mismatch", if
+      // the etag on server has changed.
+      client.updateTableParameters(
+          tableName, buildTableParameters(newMetadataLocation, metadata), etag);
+    } catch (AbortedException e) {
+      if (e.getMessage().toLowerCase().contains("etag mismatch")) {
+        throw new CommitFailedException(
+            "Updating table failed due to conflict updates (etag mismatch)");
+      }
+    }
+  }
+
+  private Table makeNewTable(TableMetadata metadata, String metadataFileLocation) {
+    Table.Builder builder = Table.newBuilder().setType(Table.Type.HIVE);
+    builder
+        .getHiveOptionsBuilder()
+        .setTableType("EXTERNAL_TABLE")
+        .setStorageDescriptor(
+            StorageDescriptor.newBuilder()
+                .setLocationUri(metadata.location())
+                .setInputFormat("org.apache.hadoop.mapred.FileInputFormat")
+                .setOutputFormat("org.apache.hadoop.mapred.FileOutputFormat")
```

Comment:
Do you want to mirror Hive properties so that Hive can read these tables when using the same metastore?

Reply:
Yes, we hope a HMS proxy can return Iceberg tables from BigLake API as they were from HMS, so engines supporting HMS can use BigLake. In the BigLake API, such tables have type Hive. In future we may add Iceberg type and store only metadata needed for Iceberg.